src/common/system.rs

```diff
@@ -324,40 +324,46 @@ pub(crate) fn fuser<P: AsRef<Path>>(
+                                        if is_lnk(&stat_info) {
+                                            let link_data = read_link(curr_path.as_path())
+                                                .upstream_with_context(&format!(
+                                                    "Failed to read link '{}'",
+                                                    curr_path.display()
+                                                ))?;
+                                            debug!(
+                                                "looking at pid: {}, fd {}, file: '{}' -> '{}'",
                                                 curr_pid,
-                                                io::Error::last_os_error()
+                                                curr_fd,
+                                                curr_path.display(),
+                                                link_data.display()
                                             );
+
+                                            if link_data.starts_with(path.as_ref()) {
+                                                debug!("sending signal {} to {}", signal, curr_pid,);
+                                                if unsafe { libc::kill(curr_pid, signal) } != 0 {
+                                                    warn!(
+                                                        "Failed to send signal {} to pid {}, error: {}",
+                                                        signal,
+                                                        curr_pid,
+                                                        io::Error::last_os_error()
+                                                    );
+                                                } else {
+                                                    sent_signals.push(curr_pid);
+                                                }
+                                                break;
+                                            }
                                         } else {
-                                            sent_signals.push(curr_pid);
+                                            return Err(Error::with_context(
+                                                ErrorKind::InvState,
+                                                &format!("file '{}' is not a link", curr_path.display()),
+                                            ));
                                         }
-                                        break;
                                     }
-                                } else {
-                                    return Err(Error::with_context(
-                                        ErrorKind::InvState,
-                                        &format!("file '{}' is not a link", curr_path.display()),
-                                    ));
+                                    Err(_) => {
+                                        debug!("continue after lstat error");
```

Comment:
This whole set of changes in the `fuser()` function seems to be "just" to allow continuing after an `lstat()` error. Do you know why this is necessary with balenaOS-to-balenaOS migrations but wasn't necessary before? We should document the reason.

Reply:
@lmbarros, @acostach See the Fibery page _Experiment: Takeover MVP_, section _Stage 2_. On the NUC conversion, takeover was unable to find a file in the proc filesystem. I could not find it manually either, so I figured it was not a significant issue. I never got to the level of understanding _why_ the file was missing. So this change essentially just ignores the error. The Fibery page has more ideas on the cause. Have either of you seen this error in the logs during a migration?

Reply:
Thanks, @kb2ma ! I have seen this error a couple of times. I actually moved the [printing of the process table](https://github.com/balena-os/takeover/pull/54/commits/513810c7b31cd865eed908b096ae79b43717ce2c) to a point before the error happens to help with troubleshooting it. (But after that I didn't look further.)

My best guess was that this was a racing issue: if a process exits between the moments we:

1. call `read_dir()` to get the list of entries under `/proc`
2. try to access that process' `/proc/PID/fd/...`

then we'll get this error.

I don't know if this is the only possible way to get this error, but this looks like one possible way at least. (I might be missing something, too... this code is tricky.)

Not very sure what to do here, @acostach , but "by default" I prefer to not merge fixes until we reasonably certain of what the issue really is.

Reply:
ack, marking as resolved